libraries/ms-common/src/x86/can_hw.c

```diff
@@ -1,27 +1,202 @@
 #include "can_hw.h"
+#include <fcntl.h>
+#include <linux/can.h>
+#include <linux/can/raw.h>
+#include <net/if.h>
+#include <poll.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include "fifo.h"
+#include "log.h"
+
+#define CAN_HW_DEV_INTERFACE "vcan0"
+#define CAN_HW_MAX_FILTERS 14
+#define CAN_HW_TX_FIFO_LEN 8
+
+typedef struct CANHwEventHandler {
+  CANHwEventHandlerCb callback;
+  void *context;
+} CANHwEventHandler;
+
+typedef struct CANHwSocketData {
+  int can_fd;
+  struct can_frame rx_frame;
+  Fifo tx_fifo;
+  struct can_frame tx_frames[CAN_HW_TX_FIFO_LEN];
+  struct can_filter filters[CAN_HW_MAX_FILTERS];
+  size_t num_filters;
+  CANHwEventHandler handlers[NUM_CAN_HW_EVENTS];
+  uint32_t delay_us;
+} CANHwSocketData;
+
+static pthread_t s_rx_pthread_id;
+static pthread_t s_tx_pthread_id;
+static pthread_mutex_t s_tx_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t s_tx_cond = PTHREAD_COND_INITIALIZER;
+
+static CANHwSocketData s_socket_data = {.can_fd = -1 };
+
+static uint32_t prv_get_delay(CANHwBitrate bitrate) {
+  uint32_t delay_us[NUM_CAN_HW_BITRATES] = {
+    1000,  // 125 kbps
+    500,   // 250 kbps
+    250,   // 500 kbps
+    125,   // 1 mbps
+  };
+
+  return delay_us[bitrate];
+}
+
+static void *prv_rx_thread(void *arg) {
+  while (true) {
+    int bytes = read(s_socket_data.can_fd, &s_socket_data.rx_frame, sizeof(s_socket_data.rx_frame));
+
+    if (s_socket_data.handlers[CAN_HW_EVENT_MSG_RX].callback != NULL) {
+      s_socket_data.handlers[CAN_HW_EVENT_MSG_RX].callback(
+          s_socket_data.handlers[CAN_HW_EVENT_TX_READY].context);
+    }
+
+    // Limit how often we can receive messages to simulate bus speed
+    usleep(s_socket_data.delay_us);
+  }
+
+  return NULL;
+}
+
+static void *prv_tx_thread(void *arg) {
+  struct can_frame frame = { 0 };
+
+  while (true) {
+    pthread_mutex_lock(&s_tx_mutex);
+    while (fifo_size(&s_socket_data.tx_fifo) == 0) {
+      pthread_cond_wait(&s_tx_cond, &s_tx_mutex);
+    }
+
+    fifo_pop(&s_socket_data.tx_fifo, &frame);
+    pthread_mutex_unlock(&s_tx_mutex);
+
+    int bytes = write(s_socket_data.can_fd, &frame, sizeof(frame));
+
+    // Delay to simulate bus speed
+    usleep(s_socket_data.delay_us);
+
+    if (s_socket_data.handlers[CAN_HW_EVENT_TX_READY].callback != NULL) {
+      s_socket_data.handlers[CAN_HW_EVENT_TX_READY].callback(
+          s_socket_data.handlers[CAN_HW_EVENT_TX_READY].context);
+    }
+  }
+
+  return NULL;
+}
 
 StatusCode can_hw_init(const CANHwSettings *settings) {
-  return status_code(STATUS_CODE_UNIMPLEMENTED);
+  if (s_socket_data.can_fd != -1) {
+    // Reinit everything
+    // This should eventually be done properly with signals
```

Comment:
This is causing some tsan sadness, but seems to work alright in our tests. Probably not the safest thing, but we'll never encounter it in application code.